openmc/model/surface_composite.py

```diff
@@ -0,0 +1,394 @@
+
+    def evaluate(self, point):
+        raise NotImplementedError('Composite surfaces do not have a surface equation.')
+
+    def _get_base_coeffs(self):
+        raise NotImplementedError('Composite surfaces do not have base coefficients.')
+
+    def translate(self, vector, inplace=False):
+        surf = self if inplace else copy(self)
+        for name in self._surface_names:
+            s = getattr(surf, name)
+            setattr(surf, name, s.translate(vector, inplace))
+        return surf
+
+    def rotate(self, rotation, pivot=(0., 0., 0.), order='xyz', inplace=False):
+        surf = copy(self)
+        for name in self._surface_names:
+            s = getattr(surf, name)
+            setattr(surf, name, s.rotate(rotation, pivot, order, inplace))
+        return surf
+
+    @property
+    def boundary_type(self):
+        return getattr(self, self._surface_names[0]).boundary_type
+
+    @boundary_type.setter
+    def boundary_type(self, boundary_type):
+        # Set boundary type on underlying surfaces, but not for ambiguity plane
+        # on one-sided cones
+        for name in self._surface_names:
+            if name != 'plane':
+                getattr(self, name).boundary_type = boundary_type
+
+    def __repr__(self):
+        return "<{} at 0x{:x}>".format(type(self).__name__, id(self))
+
+    @property
+    @abstractmethod
+    def _surface_names(self):
+        """Iterable of attribute names corresponding to underlying surfaces."""
+
+    @abstractmethod
+    def __pos__(self):
+        """Return the positive half-space of the composite surface."""
+
+    @abstractmethod
+    def __neg__(self):
+        """Return the negative half-space of the composite surface."""
+
+
+class RightCircularCylinder(CompositeSurface):
+    """Right circular cylinder composite surface
+
+    A right circular cylinder is composed of a cylinder and two planar surface
+    perpendicular to the axis of the cylinder. This class acts as a proper
+    surface, meaning that unary `+` and `-` operators applied to it will produce
+    a half-space. The negative side is defined to be the region inside of the
+    right circular cylinder.
+
+    .. versionadded:: 0.12
+
+    Parameters
+    ----------
+    center_base : iterable of float
+        Cartesian coordinate of the center of the base of the cylinder
+    height : float
+        Height of the cylinder
+    radius : float
+        Radius of the cylinder
+    axis : {'x', 'y', 'z'}
+        Axis of the cylinder
+    **kwargs
+        Keyword arguments passed to underlying cylinder and plane classes
+
+    Attributes
+    ----------
+    cyl : openmc.Cylinder
+        Underlying cylinder surface
+    bottom : openmc.Plane
+        Bottom planar surface of the cylinder
+    top : openmc.Plane
+        Top planar surface of the cylinder
+
+    """
+    _surface_names = ('cyl', 'bottom', 'top')
+
+    def __init__(self, center_base, height, radius, axis='z', **kwargs):
```

Comment:
Do you want to allow empty composite surfaces ? Like a cylinder with a negative height, or a parallelepiped with xmax < xmin

Reply:
Thanks for pointing that out -- I'll add some checks to disallow that